File: c++/2239.cpp
#include <iostream>
#include <string>

using namespace std;

int sudoku[9][9];
bool row[9][10];
bool col[9][10];
bool box[9][10];
bool endFlag;

void printArr() {
  for (int i = 0; i < 9; i++) {
    for (int j = 0; j < 9; j++) {
      cout << sudoku[i][j];
    }
    cout << '\n';
  }
}

void dfs(int cnt) {
  int y = cnt / 9;
  int x = cnt % 9;

  if (endFlag) return;

  if (cnt == 81) {
    printArr();
    endFlag = true;
    return;
  }

  if (sudoku[y][x] == 0) {
    for (int k = 1; k <= 9; k++) {
      int boxLoc = y / 3 * 3 + x / 3;
      if (!row[y][k] && !col[x][k] && !box[boxLoc][k]) {
        sudoku[y][x] = k;
        row[y][k] = true;
        col[x][k] = true;
        box[boxLoc][k] = true;
        dfs(cnt + 1);
        sudoku[y][x] = 0;
        row[y][k] = false;
        col[x][k] = false;
        box[boxLoc][k] = false;
      }
    }
  } else {
    dfs(cnt + 1);
  }
}

int main() {
  freopen("example.txt", "r", stdin);

  ios_base::sync_with_stdio(0);
  cin.tie(0);

  for (int i = 0; i < 9; i++) {
    string s;
    cin >> s;
    for (int j = 0; j < 9; j++) {
      int num = s[j] - '0';
      sudoku[i][j] = num;
      row[i][num] = true;
      col[j][num] = true;
      box[i / 3 * 3 + j / 3][num] = true;
    }
  }

  dfs(0);

  return 0;
}

File: js/2239.js
// #1
// const fs = require('fs');

// const row = Array(9)
//   .fill()
//   .map(() => Array(10).fill(false));
// const col = Array(9)
//   .fill()
//   .map(() => Array(10).fill(false));
// const box = Array(9)
//   .fill()
//   .map(() => Array(10).fill(false));

// let emptyNums = 0;
// let end = false;

// const sudoku = fs
//   .readFileSync('/dev/stdin')
//   .toString()
//   .trim()
//   .split('\n')
//   .map((e, i) =>
//     e.split('').map((v, j) => {
//       if (+v === 0) emptyNums++;
//       row[i][+v] = true;
//       col[j][+v] = true;
//       box[3 * Math.floor(i / 3) + Math.floor(j / 3)][+v] = true;
//       return +v;
//     })
//   );

// dfs(0);

// function dfs(cnt) {
//   if (end) return;

//   if (cnt === emptyNums) {
//     console.log(sudoku.map((e) => e.join('')).join('\n'));
//     end = true;
//     return;
//   }

//   for (let i = 0; i < 9; i++) {
//     for (let j = 0; j < 9; j++) {
//       if (sudoku[i][j] === 0) {
//         for (let k = 1; k <= 9; k++) {
//           const boxLoc = 3 * Math.floor(i / 3) + Math.floor(j / 3);
//           if (!row[i][k] && !col[j][k] && !box[boxLoc][k]) {
//             sudoku[i][j] = k;
//             row[i][k] = true;
//             col[j][k] = true;
//             box[boxLoc][k] = true;
//             dfs(cnt + 1);
//             sudoku[i][j] = 0;
//             row[i][k] = false;
//             col[j][k] = false;
//             box[boxLoc][k] = false;
//           }
//           if (k === 9) return;
//         }
//       }
//     }
//   }
// }

// #2 better solution
const fs = require('fs');

const row = Array(9)
  .fill()
  .map(() => Array(10).fill(false));
const col = Array(9)
  .fill()
  .map(() => Array(10).fill(false));
const box = Array(9)
  .fill()
  .map(() => Array(10).fill(false));

let end = false;
const sudoku = fs
  .readFileSync('example.txt')
  .toString()
  .trim()
  .split('\n')
  .map((e, i) =>
    e.split('').map((v, j) => {
      row[i][+v] = true;
      col[j][+v] = true;
      box[3 * Math.floor(i / 3) + Math.floor(j / 3)][+v] = true;
      return +v;
    })
  );

dfs(0);

function dfs(cnt) {
  const y = Math.floor(cnt / 9);
  const x = cnt % 9;

  if (end) return;

  if (cnt === 81) {
    console.log(sudoku.map((e) => e.join('')).join('\n'));
    end = true;
    return;
  }

  if (sudoku[y][x] === 0) {
    for (let k = 1; k <= 9; k++) {
      const boxLoc = 3 * Math.floor(y / 3) + Math.floor(x / 3);
      if (!row[y][k] && !col[x][k] && !box[boxLoc][k]) {
        sudoku[y][x] = k;
        row[y][k] = true;
        col[x][k] = true;
        box[boxLoc][k] = true;
        dfs(cnt + 1);
        sudoku[y][x] = 0;
        row[y][k] = false;
        col[x][k] = false;
        box[boxLoc][k] = false;
      }
    }
  } else {
    dfs(cnt + 1);
  }
}
